### DISCUSSION TITLE
ES2024に対応するIssueへのContributorを募集しています

### DISCUSSION THREAD

**Discussion by azu** [asciidwango/js-primer#1727] [General]
次のメジャーアップデート(v6)では、2024年6月末ごろにリリースされる予定のECMAScript 2024に対応する予定です。
そのため、ES2024の対応とユースケースであるNode.jsに関するアップデートをしていくマイルストーンを切りました。

- [v6(ES2024) Milestone](https://github.com/asciidwango/js-primer/milestone/5)

[ES2024に対応するマイルストーン](https://github.com/asciidwango/js-primer/milestone/5)をざっとみると、大体ざっくり換算12-20日分ぐらいのタスクがあります。

<img width="733" alt="image" src="https://github.com/asciidwango/js-primer/assets/19714/96647865-e1b0-4538-b100-c4217d2421a1">

大きく分けて次の3つのグループに分かれています。

- ES2024の対応
- Stage 2.7の追加
- Node.jsのユースケースの更新

### ES2024の対応

ES2024ではいくつか機能が増えているので、それに対応するアップデートを行います。
[ECMAScript 2024の対応 · Issue #1706](https://github.com/asciidwango/js-primer/issues/1706)というIssueに、ES2024の対応についてのタスクをまとめています。

具体的に対応したいIssueは次の3つになっています。

- [ES2024: array-grouping · Issue #1716](https://github.com/asciidwango/js-primer/issues/1716)
- [ES2024: Well-Formed Unicode Strings · Issue #1715](https://github.com/asciidwango/js-primer/issues/1715)
- [ES2022: Error Cause · Issue #1714](https://github.com/asciidwango/js-primer/issues/1714)

Array Groupingは恐らく対応が必須ですが、他の2つは可能なら対応する形になると思います。
これはjsprimerはリファレンスを作ることは、[目的ではないこと](https://jsprimer.net/intro/#do-not)だからです。
なので、含めるべきか or 含めても不自然ではないかを調べるという感じのIssueです。
(あった方がいいかをちゃんと判断する)

### Stgae 2.7の追加

ECMAScriptの仕様策定プロセス自体が変更されているので、それを解説する[ECMAScript · JavaScript Primer #jsprimer](https://jsprimer.net/basic/ecmascript/)を更新する予定です。

- [Stage 2.7を追加する · Issue #1713](https://github.com/asciidwango/js-primer/issues/1713)

多分これは自分がやる気はしていますが、やりたい方いたらコメントください。

### Node.jsのユースケースの更新

[Node.jsでCLIアプリ · JavaScript Primer #jsprimer](https://jsprimer.net/use-case/nodecli/)の章を全体的にアップデートする予定です。

[Meta: Node.jsのユースケースの更新 · Issue #1719](https://github.com/asciidwango/js-primer/issues/1719)というMeta Issueに依存関係をまとめています。
最近は、今まではnpmパッケージを使わないといけなかったものが、Node.jsの標準機能になったりしているので、それに対応する形のリファクタリングになります。

- [Node.js 20 LTSへアップデートする · Issue #1718](https://github.com/asciidwango/js-primer/issues/1718)
- [mochaを `node:test` に変更する · Issue #1717](https://github.com/asciidwango/js-primer/issues/1717)
- [utils.parseArgsの利用法を追記 · Issue #1698](https://github.com/asciidwango/js-primer/issues/1698)
- [markedjsのアップデート · Issue #1694](https://github.com/asciidwango/js-primer/issues/1694)

こちらは、サンプルコードのアップデートやそれにあわせた文章の変更が中心になります。
そのため、基本的に文章の意味合いを変えるような変更は少なくて、リファクタリング的な変更が中心になります。

## jsprimerのContributorを募集しています

今回のjsprimerのES2024の対応を手伝ってくれるContributorを募集しています！

先ほども書いていましたが、おおよそざっくり12-20日ぐらいのタスクがあります。
jsprimerは文章のプロジェクトですが、[textlint](https://textlint.github.io/)でのチェックやテストも多いので、必要以上に恐れずに手伝ってくれると嬉しいです。

Node.jsのユースケースの更新は、Node.jsの新しい機能を使うように変更するというのが主なので、そこに興味がある方は特に歓迎です。
こちらは、文章の意味合いはあまり変える必要はないので(コードが変わるのでそれに合わせた変更や流れの調整は必要です)、比較的やりやすいかと思います。

ただし、Node.jsはタスク間の依存関係がちょっとややこしいところが複数人だとちょっとやりにくいかもれないです。
(文章を自然に読めるようにするために未知のことに依存させないという方針なので、文章の並び替えが必要になります)

- [Meta: Node.jsのユースケースの更新 · Issue #1719 · asciidwango/js-primer](https://github.com/asciidwango/js-primer/issues/1719)

もし、やってみたい方とかがいれば、一度認識合わせのためにMTGしたり、レビューとかは当然やるのでコメントなどでお知らせください。

(もちろん [他のIssue](https://github.com/asciidwango/js-primer/issues)へのContirbutionはいつでも歓迎しています！)

**Comment by koralle** [comment]
> 今回のjsprimerのES2024の対応を手伝ってくれるContributorを募集しています！

初めまして！koralle（こーらる）です。
こちらぜひお手伝いしたいです！

**Comment by yossydev** [comment]
初めまして！ぜひお手伝いさせていただきたいです！

**Comment by azu** [comment]
@koralle @yossydev ありがとうございます！
このissueやりたい！というものはあったりしますか？（逆にこのIssueよくわからないとかでも大丈夫です）

特になければ、とりあえず自分が一回ざっと話してみて、興味あるもの選んでもらうのがいいかなと思っていますがどうでしょうか？
ここでやってたようなVideo Meetingするイメージです。
https://github.com/asciidwango/js-primer/tree/master/meetings

**Reply by yossydev** [reply]
自分はNode.jsのユースケースの更新がとても興味あるので可能であればそちらをやりたいです！
[Node.js 20 LTSへアップデートする](https://github.com/asciidwango/js-primer/discussions/1727#discussioncomment-8866378)からかと思いましたが、こちらはすでに完了しているんですかね...？？👀

**Reply by azu** [reply]
了解です。

> https://github.com/asciidwango/js-primer/discussions/1727#discussioncomment-8866378からかと思いましたが、こちらはすでに完了しているんですかね...？？👀

そうですね20 LTSへのアップデートは #1718 完了してます。
Node.jsのやつは、npm installを使うタイミングが後ろにずれる(`node:`パッケージに変更する)都合的に、章の中身を入れ替える感じです。

3つissueありますが、互いに依存しちゃってると思うので順番にやらないとconflictするとは思います。
一応、次のmeta Issueにまとめてあります。

- #1719 

多分ややこしいのは https://github.com/asciidwango/js-primer/issues/1698 に書いてあるこの辺の話(commanderは使わなくなるので`npm install`を使うタイミングが後ろにずれる)なので、この辺のissueだけを別issueに切り出すとかやるとかもアリかもしれないですね

```
- [ ] mpm initとcommonjsの件を [コマンドライン引数を処理する](https://jsprimer.net/use-case/nodecli/argument-parse/) → [Node.jsでHello World](https://jsprimer.net/use-case/nodecli/helloworld/) に移動する
  - プロジェクトのセットアップの話としてnpm initする
```

**Reply by koralle** [reply]
@azu 
可能であればES2024対応やってみたいです！

**Reply by yossydev** [reply]
@azu 
ご丁寧にありがとうございます！！

> 多分ややこしいのは https://github.com/asciidwango/js-primer/issues/1698 に書いてあるこの辺の話(commanderは使わなくなるのでnpm installを使うタイミングが後ろにずれる)なので、この辺のissueだけを別issueに切り出すとかやるとかもアリかもしれないですね

https://github.com/asciidwango/js-primer/issues/1731 上記に関してissueを切っておきました！こちらそのまま自分が進めて問題ないでしょうか...？？👀

**Reply by koralle** [reply]
> @azu 
> 可能であればES2024対応やってみたいです！

これは可能でしょうか...??
https://github.com/asciidwango/js-primer/issues/1716

**Reply by azu** [reply]
@koralle はい大丈夫です！
Object.groupBy の方はreduceの後ろあたりにセクションを追加して書くのがシンプルなのかなと思ってます。
Map.groupBy はちょっとどうするか悩みどころです

**Reply by koralle** [reply]
@azu 
> Object.groupBy の方はreduceの後ろあたりにセクションを追加して書くのがシンプルなのかなと思ってます。

レスポンス遅くなってしまいすみません。
これってもう早速手をつけ始めてしまって問題ないでしょうか？

**Reply by azu** [reply]
@koralle 他のものには特に依存してないので、大丈夫ですよ

**Reply by koralle** [reply]
ありがとうございます。
では進めていきます。

**Comment by SotaYamaguchi** [comment]
はじめましてです！
私も是非参加させて頂きたいです！
着手するタスクについては、［Node.jsのユースケースの更新］のどれかあたりから取れればと考えていますmm

**Comment by windchime-yk** [comment]
はじめまして！ WhyKと申します。
是非お手伝いしたいです！
個人的には、『Node.jsのユースケースの更新』の『mochaを`node:test`に変更する』に興味があります！

**Comment by jp-knj** [comment]
ES2024 対応に興味があります。参加させていただけないでしょうか？

こちらの issue をやってみたいなと考えています。
- https://github.com/asciidwango/js-primer/issues/1698

**Comment by azu** [comment]
@yossydev @SotaYamaguchi @windchime-yk @jp-knj
Node.jsの方に興味に持っていただきありがとうございます。
多分一斉にやると間違いなくConflictすると思います。

- https://github.com/asciidwango/js-primer/issues/1719

次の3つはおそらく順番にやらないと、作業途中(それぞれのIssueが解決されてる段階でも読める状態にしたい)の文章が意味わからない感じになるかなと思ってます。

- [ ] https://github.com/asciidwango/js-primer/issues/1731 
  - [ ] #1698 
    - [ ] #1717 

どうやるのが最適ですかね?

1. それぞれのIssueにAssignしていって、互いにreviewしながら一つずつ進める
  - 依存してるissueが解決されるまで待ち時間がある
2. 一度 tempなbranchを用意してそこにマージしていって、まとめてマージする
  - conflictが起きること自体は解決されない

(Node.jsの変更は順番を変えたりが必要で、全体像を見ながら進めた方が読んでて違和感ないものができると思うので、AssignされたIssue以外も見ながらやるのが良いかなーとは思っています)

## まとめ(追記)

- 変更箇所とIssueの対応: https://github.com/asciidwango/js-primer/discussions/1727#discussioncomment-8883661
- 一時的にまとめるブランチ: `v6-node-updates` へPRを出す

**Reply by azu** [reply]
ちょっと一回アウトラインレベルで変更の順序整理してみます。
https://github.com/asciidwango/js-primer/issues/1731 の移動は実はしなくても問題ないかもしれない気がしてきました。(`node:utils` をESMとしてimportするためには、ここで npm init --yes && type=moduleしても不自然ではないかもしれない。)

**Reply by azu** [reply]
アウトラインを整理して、Issueごとに変更する場所を可視化してみました。

[![image](https://github.com/asciidwango/js-primer/assets/19714/2b4d0073-00c3-4b5a-81cd-a28b8728ca53)](https://efcl.notion.site/Node-js-7ff26cae41da41c194abe1ac3b698a28?pvs=74)

- https://efcl.notion.site/Node-js-7ff26cae41da41c194abe1ac3b698a28?pvs=74

それぞれのIssueで変更する箇所を 色で分けています。

- 🟢 https://github.com/asciidwango/js-primer/issues/1698
    - commander → `node:util` の `parseArgs` を使うように変更
        - node: を初めて使うタイミングが早まるので、 `node: の説明` を fsモジュールから移動させる
    - `npm install` をする初めてのタイミングが `marked` までずれてしまうので、 [markedパッケージをインストール](https://www.notion.so/marked-cd4dc59145124959bd9a8efdcd632d43?pvs=21) を新しく作って、`npm install`の説明をここに移動させる
        - 中身は大体 commander のインストールとほぼ同じ
- 🟣 https://github.com/asciidwango/js-primer/issues/1717
    - `mocha` → `node:test` に変更する
    - 他の変更とはあまり関係なくできるかも
        - `package.json` のconflictだけは起きるけど、文章の変更は他とは独立してる
    - ~~ただしマージするタイミングは、 https://github.com/asciidwango/js-primer/issues/1698 の後になる~~ (文章だけのマージなら独立してそうだった)
- 🟠 https://github.com/asciidwango/js-primer/issues/1694
    - 多分変更点は一番少ないはず
        - 文章的な変更はほとんどないはず? (GFMのデフォルトとかの調査は必要)
    - ほかの作業の重なる部分も変更が必要なので、一番最後にやった方が良さそう
        - [markedパッケージをインストール](https://www.notion.so/marked-cd4dc59145124959bd9a8efdcd632d43?pvs=21)が増えてからバージョン部分だけいじる
        - [ユニットテストを記述する](https://jsprimer.net/use-case/nodecli/refactor-and-unittest/#write-unit-test)

この2つは、並行してやってもそこまでConflictしない感じはします(package.jsonだけ多分Conflictしますが、installしなおせば治る)。マージするタイミングは #1698 → #1717 になります。

- https://github.com/asciidwango/js-primer/issues/1698
- https://github.com/asciidwango/js-primer/issues/1717

markedのアップデートは、他の2つが終わってからが楽かなと思います。

- https://github.com/asciidwango/js-primer/issues/1694

----

- https://github.com/asciidwango/js-primer/issues/1698

こちらが変更範囲が多いので、 @yossydev さん と @jp-knj さんにお願いしたいです。
(ちょっとどう分担するかのアイデアは出てこなかったです。すみません)

- https://github.com/asciidwango/js-primer/issues/1717

こちら @windchime-yk さんにお願いしたいです。

- https://github.com/asciidwango/js-primer/issues/1694 

こちら @SotaYamaguchi さんにお願いしたいです。
ちょっとアップデートでどういう変更が必要かが見えてないので、アップデートしてみて文章やテストを変更しないといけない箇所を調べるところからやるのが良いと思います。


cc @lacolaco

**Reply by jp-knj** [reply]
@azu 

> どうやるのが最適ですかね?
> 
> 2. 一度 tempなbranchを用意してそこにマージしていって、まとめてマージする

私個人の感想なのですが、2 で良さそうだなって思ってます。
すべての変更が一つのブランチに集約されるため、全体像を把握しやすそうです。
互いの作業がどのように連携しているかを理解しそうだなって思っちゃいました。

私が担当することになる issue は、どのみち at yossydev さんとコンフリクトしやすくなりそうなので、
コンフリクトを承知で進めてていきたいなと考えてます。多くの変更が一つのブランチに統合されると、コンフリクトが複雑になりそうなので、そこらへん、気をつけて進めていけば、解決できそうかなだなと捉えてます。（希望的観測ですが）

> (ちょっとどう分担するかのアイデアは出てこなかったです。すみません)

いえ、ありがとうございます！

**Reply by azu** [reply]
一応 `v6-node-updates` というブランチを切っておきました
PRを出すときに、PRの向き先を `v6-node-updates` にしておいて、 `v6-node-updates` を `master` にマージし直す感じですかね。

📝 特にすぐマージしても問題ないやつ(ES2024とかマージ順は特に関係ないもの)は直接`master`にPR出して問題ないです。

**Comment by himanoa** [comment]
こんにちは！是非お手伝いさせていただきたいです！
興味あるIssueは https://github.com/asciidwango/js-primer/issues/1714 です

**Reply by azu** [reply]
@himanoa ありがとうございます！ OKです！

**Comment by azu** [comment]
https://github.com/asciidwango/js-primer/pull/1743
Stage 2.7は追加済み

**Comment by azu** [comment]
ES2024がEcma GA(総会)で承認されたの正式リリースになりました
https://github.com/tc39/ecma262/releases/tag/es2024
https://2ality.com/2024/06/ecmascript-2024.html

**Reply by azu** [reply]
今月後半ぐらいまで書く時間は取れない気がするので、リリースは来月ぐらいにできると良さそう。
レビュー自体はいつでもできるので、いつでも投げてください

**Comment by azu** [comment]
@himanoa @windchime-yk お待たせしました。

Open Collectiveでの経費申請の方法を整えたので、もし申請する場合は次の手順で申請できます

- https://github.com/asciidwango/js-primer/blob/master/CONTRIBUTING_EXPENSE.md

申請方法の詳細は、上記のドキュメントを参考にしてください。
それぞれ 2point換算(現時点だと`$78.9`)で良いと思います。

ちょっと実際のcommitと日付が空いてるので、それぞれ日付を範囲にしてしまうのが良い気がします。

- Expense title: `Development and maintenance (2024-03 - 2024-08)`
- Description: `Contribute to jsprimer (2024-03 - 2024-08)`
- Add notes: `Commits: https://github.com/asciidwango/js-primer/commits?author=<account>&since=2024-03-01&until=2024-09-01`

ちょっとまだ実際にやってみてはないので、実際にどのタイミングで支払われるかがわかってないですが、何かあればコメントか `azuciao@gmail.com` にメールしてください。

**Reply by windchime-yk** [reply]
@azu 
手順のご連携ありがとうございます！
申請しようと思っているのですが、co-commitだとAdd notesのURLで`<account>`を自分のものにしても表示されないようです。
https://github.com/asciidwango/js-primer/commits?since=2024-03-01&until=2024-09-01&author=windchime-yk
userをazuさんで絞り込むとco-committerとして表示されてはいるので、そちらで申請しようと考えていますが問題ないでしょうか？

**Reply by azu** [reply]
あー、なるほど。一度mainじゃないブランチを経由したからな気がしますね。
( https://github.com/asciidwango/js-primer/graphs/contributors こちらにはちゃんと出てるんですが、commitsには出ないんですね)

コミットとPRのURLをそのまま書いてしまうのが良いと思います

- https://github.com/asciidwango/js-primer/commit/2f99346124f08d9dea25be5a978b94819c73e4d4
- https://github.com/asciidwango/js-primer/pull/1737

**Reply by windchime-yk** [reply]
なるほど、ありがとうございます！
CommitとPRのURLで申請させていただきます！

**Reply by azu** [reply]
ありがとうございます。ApproveしたのでOpen Source Collective側の承認が通ったら支払われると思います。(2週間ごととかにまとめてされるみたいな話をどこかで見た気がします)

**Comment by azu** [comment]
https://github.com/asciidwango/js-primer/releases/tag/v6.0.0
リリースしたのでcloseします
